Loading and Preprocessing

In [52]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.svm import SVR
from sklearn.metrics import mean_squared_error, mean_absolute_error,r2_score

In [6]:
housing = fetch_california_housing()
df = pd.DataFrame(housing.data,columns = housing.feature_names)
df.isnull().sum()
df["MedHouseVal"] = housing.target
df.head()


,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,MedHouseVal
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422


Checking for missing value is an important step in data preprocessing. Because null input values may result in wrong evaluvation.

Separate Features and Target

In [7]:
X = df.drop("MedHouseVal", axis=1)

y = df["MedHouseVal"]

Separating independent and depent variables for model evaluvation

Split the Dataset

In [8]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


Dataset is split for traing and testing data. Training data is set as 80% and test data is 20%

In [11]:
scaler = StandardScaler()
scaled_x_train = scaler.fit_transform(X_train)
scaled_x_test = scaler.transform(X_test)

Standardization transforms the features approximately so that they have:

mean = 0
standard deviation = 1

Linear Regression :It is suitable as a baseline model because it is simple and helps us understand how well a linear relationship can explain house prices.

In [12]:
linear_reg = LinearRegression()
linear_reg.fit(scaled_x_train, y_train)
y_pred = linear_reg.predict(scaled_x_test)

Decision Tree Regressor :It can capture non-linear relationships, which is useful because house prices are unlikely to have a purely linear relationship with all features.

In [25]:
dt = DecisionTreeRegressor(random_state=42)

dt.fit(scaled_x_train, y_train)

y_pred_dt = dt.predict(scaled_x_test)

Random Forest Regressor:This generally makes the model more robust than a single Decision Tree and can capture complex relationships in the California housing data.

In [26]:
rf = RandomForestRegressor(
    n_estimators=100,
    random_state=42,
    n_jobs=-1
)

rf.fit(scaled_x_train, y_train)

y_pred_rf = rf.predict(scaled_x_test)

Gradient Boosting Regressor:Gradient Boosting builds trees sequentially.

Each new tree attempts to improve the errors made by the previous trees.

In [29]:
gb = GradientBoostingRegressor(
    n_estimators=100,
    random_state=42
)

gb.fit(scaled_x_train, y_train)

y_pred_gb = gb.predict(scaled_x_test)

Support Vector Regressor: SVR tries to find a function that predicts the target while allowing a certain amount of error.

SVR can model non-linear relationships when an appropriate kernel is used.

In [30]:
svr = SVR(kernel="rbf")

svr.fit(scaled_x_train, y_train)

y_pred_svr = svr.predict(scaled_x_test)

Evaluvate the models of Linear Regression

In [53]:
mse = mean_squared_error(y_test, y_pred)
print("mse :", mse)
mae = mean_absolute_error(y_test, y_pred)
print("mae :", mae)
r2_lr = r2_score(y_test, y_pred)
print("r2_score :",r2_lr)

mse : 0.5558915986952442
mae : 0.5332001304956565
r2_score : 0.575787706032451


Linear Regression: The model achieved an MSE of 0.5558, an MAE of 0.5332, and an R² score of 0.5757. The R² score indicates that the model explains approximately 57.57% of the variation in median house values.

Evaluation of Decision Tree Regressor

In [54]:
mse = mean_squared_error(y_test, y_pred_dt)
print("mse :", mse)
mae = mean_absolute_error(y_test, y_pred_dt)
print("mae :", mae)
r2_dt = r2_score(y_test, y_pred_dt)
print("r2_score :", r2_dt)

mse : 0.49396854311945243
mae : 0.45390448401162786
r2_score : 0.6230424613065773


Decision Tree Regressor :The model achieved an MSE of 0.4939, an MAE of 0.4539, and an R² score of 0.6230 The R² score indicates that the model explains approximately 62.30% of the variation in median house values.

Evaluvation of Random Forest Regressor

In [55]:
mse = mean_squared_error(y_test, y_pred_rf)
print("mse :", mse)
mae = mean_absolute_error(y_test, y_pred_rf)
print("mae :", mae)
r2_rf = r2_score(y_test, y_pred_rf)
print("r2_score :", r2_rf)

mse : 0.255169737347244
mae : 0.3274252027374033
r2_score : 0.8052747336256919


Random Forest Regressor:The model achieved an MSE of 0.2551, an MAE of 0.3274, and an R² score of 0.8052 The R² score indicates that the model explains approximately 80.5% of the variation in median house values.

Evaluation of Gradient Boosting Regressor

In [56]:
mse = mean_squared_error(y_test, y_pred_gb)
print("mse :", mse)
mae = mean_absolute_error(y_test, y_pred_gb)
print("mae :", mae)
r2_gb = r2_score(y_test, y_pred_gb)
print("r2_score :",r2_gb)

mse : 0.29399901242474274
mae : 0.37165044848436773
r2_score : 0.7756433164710084


Gradient Boosting Regressor :The model achieved an MSE of 0.2939, an MAE of 0.3716, and an R² score of 0.6230 The R² score indicates that the model explains approximately 77.56% of the variation in median house values.

Evaluvation of SVR

In [57]:
mse = mean_squared_error(y_test, y_pred_svr)
print("mse :", mse)
mae = mean_absolute_error(y_test, y_pred_svr)
print("mae :", mae)
r2_svr = r2_score(y_test, y_pred_svr)
print("r2_score :",r2_svr)

mse : 0.357004031933865
mae : 0.39859907695205365
r2_score : 0.7275628923016773


SVR:The model achieved an MSE of 0.3570, an MAE of 0.3985, and an R² score of 0.7275 The R² score indicates that the model explains approximately 72.75% of the variation in median house values.

Conclusion :After comparing all five models, the model with the lowest MSE and MAE and the highest R² will be considered the best-performing model. The model with the highest errors and lowest R² will be considered the worst-performing model. RandomForestRegressor is the best model and LinearRegression is the worst model